### PR TITLE
Fix json serialization

### PR DESCRIPTION
**Pull request by arogachev**
Reverts yiisoft/var-dumper#67

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,36 +3,37 @@
 
 ## 1.2.1 under development
 
-- Enh #64: Add `asJson()` method to `\Yiisoft\VarDumper\VarDumper` (dood-)
+- Enh #64: Add `asJson()` method to `\Yiisoft\VarDumper\VarDumper` (@dood-)
+- Enh #68: Fix json serialization (@xepozz)
```
Comment (vjik) [inline CHANGELOG.md:7]:
```suggestion
- New #64, 68: Add `VarDumper::asJson()` method that provide exporting value as JSON string (@dood-, @xepozz)
```

File: src/VarDumper.php [hunk 2]
```diff
@@ -39,6 +39,9 @@
  */
 final class VarDumper
```
Comment (vjik) [inline src/VarDumper.php:40]: Need 100% test coverage for all new code.

File: src/VarDumper.php [hunk 3]
```diff
@@ -469,7 +501,17 @@ private function exportVariable($variable): string
...
     private function getObjectDescription(object $object): string
     {
-        return get_class($object) . '#' . spl_object_id($object);
+        return $this->getObjectClass($object) . '#' . $this->getObjectId($object);
+    }
+
+    private function getObjectClass(object $object): string
+    {
+        return get_class($object);
+    }
```
Comment (vjik) [inline src/VarDumper.php:510]: Looks as overengineering. You can use `get_class($object)` directly.

File: src/VarDumper.php [hunk 4]
```diff
@@ -39,6 +39,9 @@
  */
 final class VarDumper
 {
+    public const OBJECT_ID_PROPERTY = '__id__';
+    public const OBJECT_CLASS_PROPERTY = '__class__';
+    public const DEPTH_LIMIT_EXCEEDED_PROPERTY = '__depth_limit_exceeded__';
```
Comment (vjik) [inline src/VarDumper.php:44]: A potential conflict is possible. Suggest use symobls that deny to use in object property name.

File: src/VarDumper.php [hunk 5]
```diff
@@ -332,23 +335,52 @@ private function exportInternal($variable, bool $format, int $level): string
...
+                $output = [
+                    self::OBJECT_ID_PROPERTY => $objectId,
+                    self::OBJECT_CLASS_PROPERTY => $objectClass,
+                ];
+                /** @psalm-var mixed $value */
+                foreach ($objectProperties as $name => $value) {
+                    $propertyNames = explode("\0", trim((string) $name));
+                    $propertyName = end($propertyNames) ?: $name;
+                    /** @psalm-suppress MixedAssignment */
+                    $output[$propertyName] = $this->exportJson($value, $format, $depth, $level + 1);
```
Comment (vjik) [inline src/VarDumper.php:381]: With `$output[$name]` tests pass. Need test for this case.

**Review by vjik — COMMENTED** [review]
Looks good, need fix psalm errors only.

Maybe extract code for generate string and JSON from `VarDumper` class to separate classes (In separate PR)? What you think?

**Comment by xepozz** [comment]
> Maybe extract code for generate string and JSON from VarDumper class to separate classes (In separate PR)? What you think?

We may think about moving json serialisation, for example, into yiisoft/json.

**Pull request merged**